lib/FrontendTool/FrontendTool.cpp

```diff
@@ -135,6 +135,93 @@ swift::frontend::utils::escapeForMake(StringRef raw,
   return buffer.data();
 }
 
+// From the given set of frontend options and the primary input file, apply a
+// set of callbacks that enable the construction of make-style dependencies with
+// minimal duplication.
+//
+// The idea is to collate the set of known outputs in a preferred order. Each of
+// these outputs will become a target listed in the .d file, but only the first,
+// most-preferred target needs the full list of files as its dependencies.  The
+// remaining targets need only depend on that target being rebuilt to be
+// re-emitted.
+//
+// Thus, the first callback is invoked for the most preferred target - often an
+// object file or a .swiftmodule file.  The remaining callback is provided with
+// an auxiliary target as its first argument, and the main target on which
+// it should depend as its second argument.
+//
+// The final structure of the dependency file is therefore:
+//
+// /Path/To/Most/Preferred/Output/File.o:
+//   <Big Ol' List Of Files...>
+// /Path/To/Supplementary/Output/File.swiftdoc:
+//   /Path/To/Most/Preferred/Output/File.o
+// /Path/To/Supplementary/Output/File.swiftsourceinfo:
+//   /Path/To/Most/Preferred/Output/File.o
+// ...
+static void forEachMinimalMakeStyleTarget(
+    const FrontendOptions &opts,
+    const InputFile &input,
+    llvm::function_ref<void(StringRef)> mainTarget,
+    llvm::function_ref<void(StringRef, StringRef)> remainingTarget) {
+  llvm::SmallVector<const std::string *, 8> preferredOutputs;
+  llvm::StringSet<> seenTargets;
+
+  // Try to compute the main file from the action we're given.  If we're just
+  // emitting a module interface or running merge-modules, we'll pick up the
+  // module output path first from the list below.
+  if (opts.RequestedAction != FrontendOptions::ActionType::EmitModuleOnly &&
+      opts.RequestedAction != FrontendOptions::ActionType::MergeModules) {
+    if (opts.InputsAndOutputs.isWholeModule()) {
+      opts.InputsAndOutputs.forEachInputProducingAMainOutputFile([&](const InputFile &input) {
+        // Unique the file names we're pushing as targets. Because multiple
+        // source files are going to map to the same main output file.
+        if (seenTargets.insert(input.outputFilenameRef()).second) {
+            preferredOutputs.push_back(&input.outputFilenameRef());
+        }
+        // Don't early-exit.
+        return false;
+      });
+    } else {
+      preferredOutputs.push_back(&input.outputFilenameRef());
+    }
```

Comment:
I find this easier to read when the short branch, here the else case, comes first. That way I don't have to match up the brace before the "else" with the appropriate "if".